src/Microsoft.AspNetCore.Server.IntegrationTesting/Common/TestUriHelper.cs

```diff
@@ -11,39 +11,68 @@ public static class TestUriHelper
     {
         public static Uri BuildTestUri()
         {
-            return new UriBuilder("http", "localhost", GetNextPort()).Uri;
+            return BuildTestUri(null);
         }
 
         public static Uri BuildTestUri(string hint)
+        {
+            return BuildTestUri(hint, statusMessagesEnabled: false);
+        }
+
+        internal static Uri BuildTestUri(string hint, bool statusMessagesEnabled)
         {
             if (string.IsNullOrEmpty(hint))
             {
-                return BuildTestUri();
+                if (statusMessagesEnabled)
+                {
+                    // Most functional tests use this codepath and should directly bind to dynamic port "0" and scrape
+                    // the assigned port from the status message, which should be 100% reliable since the port is bound
```

Comment:
Does SelfHostDeployer already scrape the assigned port from the status message? Does this mean that any test that uses SelfHostDeployer w/ Kestrel w/o specifying a port should just work?

Reply:
SelfHostDeployer already scrapes the assigned port when status messages are enabled (which is the default for most tests).  So yes, once this change is merged and repos are updated to use this version, most tests should get more reliable with no other changes.